sys/dev/usb/uvc/uvc_drv.c

```diff
@@ -0,0 +1,2720 @@
+			continue;
+		ret = uvc_drv_parse_standard_ctrl(sc, desc, ctrl);
+		if (ret) {
+			return ret;
+		}
+	}
+
+	if (uvc_debug >= 5)
+		uvc_drv_show_ctrl(ctrl);
+	return 0;
+}
+
+static struct uvc_drv_ctrl *
+uvc_drv_init_ctrl(struct usb_interface *iface, uint8_t iface_index,
+	uint8_t iface_num)
+{
+	struct uvc_drv_ctrl *pc;
+
+	pc = malloc(sizeof(*pc), M_UVC, M_ZERO | M_WAITOK);
+	if (pc) {
+		pc->iface = iface;
+		pc->iface_index = iface_index;
+		pc->iface_num = iface_num;
+		mtx_init(&pc->mtx, "uvc ctrl lock", NULL, MTX_DEF);
+		STAILQ_INIT(&pc->entities);
+	}
+	return pc;
+}
+
+static void
+uvc_drv_destroy_ctrl(struct uvc_drv_ctrl *c)
+{
+	struct uvc_drv_entity *ent, *tmp;
+	uint8_t i = 0;
+
+	if (c) {
+		STAILQ_FOREACH_SAFE(ent, &c->entities, link, tmp) {
+			for (i = 0; i < ent->ncontrols; ++i) {
+				struct uvc_control *ctrl = &ent->controls[i];
+
+				if (!ctrl->initialized)
+					continue;
+
+				uvc_ctrl_cleanup_mappings(ctrl);
+				free(ctrl->uvc_data, M_UVC);
+			}
+			free(ent->controls, M_UVC);
+			free(ent, M_UVC);
+		}
+		free(c, M_UVC);
+	}
+}
+
+static int
+uvc_drv_probe(device_t dev)
+{
+	struct usb_attach_arg *uaa = device_get_ivars(dev);
+
+	if (uaa->usb_mode != USB_MODE_HOST)
+		return ENXIO;
+
+	if ((uaa->info.bInterfaceClass == UICLASS_VIDEO) &&
+			(uaa->info.bInterfaceSubClass == UISUBCLASS_CTRL))
+		return BUS_PROBE_SPECIFIC;
+
+	/* TODO: streaming interface should be flag as attached */
+
+	return ENXIO;
+}
+
+static int
+uvc_drv_detach(device_t self)
+{
+	struct uvc_softc *sc = device_get_softc(self);
+
+	DPRINTF("%s-%d-%p\n", __func__, __LINE__, self);
+	//delete sysfs
+	//sysfs_video_del_device(sc->udev, sc->iface);
+
+	/* destroy video hard code only one */
+	uvc_drv_destroy_video(sc->video);
+
+	/* driver cleanup */
+	uvc_drv_destroy_data(sc->data);
+	uvc_drv_destroy_ctrl(sc->ctrl);
+
+	if (sc->video) {
+		/* entry close */
+		uvc_v4l2_unreg(sc->video);
+		/* data free */
+		free(sc->video, M_UVC);
+		sc->video = NULL;
+	}
+	DPRINTF("%s-%d-%p\n", __func__, __LINE__, self);
+	return 0;
+}
+
+struct usb_vid_pid {
+	uint16_t	idVendor;
+	uint16_t	idProduct;
+};
+
+static struct usb_vid_pid usb_webcam_power_save_lut[] = {
+	{0x1bcf, 0x28c4},       // 5470 MTC Webcam
+	{0x1bcf, 0x2b98},	// 5470 MTC Webcam
+	{0x0c45, 0x671e},       // 5470 MTC Webcam
+	{0x0bda, 0x5520},	// 5470 MTC CNFHH46P27540009CCD0 IntegratedWebcamHD
+	{0x0bda, 0x5675},	// 5470 MTC AzureWave IntegratedWebcamHD
+	{0x0b0e, 0x3013},	// jabra panacast 50
+};
+
+static int
+usb_webcam_power_save(device_t dev, const char *product)
+{
+	int	i;
+	uint32_t vid, pid;
+
+	device_get_usb_vidpid(dev, &vid, &pid);
+	for (i = 0; i < nitems(usb_webcam_power_save_lut); i++) {
+		if (vid == usb_webcam_power_save_lut[i].idVendor &&
+		    pid == usb_webcam_power_save_lut[i].idProduct)
+			return 1;
+	}
+
+	if (product && !strncmp(product, "IntegratedWebcam", 16)) {
+		return 1;
+	}
+
+	return 0;
+}
+
+//import from linux kernel
```

Comment:
Not sure if it's really necessary in the comment, it could be in the commit message.